akka-http2-support/src/main/scala/akka/http/impl/engine/http2/RequestParsing.scala

```diff
@@ -69,10 +69,12 @@ private[http2] object RequestParsing {
 
           if (tlsSessionInfoHeader.isDefined) headers += tlsSessionInfoHeader.get
 
-          val entity =
-            if (subStream.data == Source.empty || contentLength == 0) HttpEntity.Empty
-            else if (contentLength > 0) HttpEntity.Default(contentType, contentLength, subStream.data)
-            else HttpEntity.Chunked.fromData(contentType, subStream.data)
+          val entity = subStream match {
+            case s if s.data == Source.empty || contentLength == 0 ⇒ HttpEntity.Empty
+            case ByteHttp2SubStream(_, data) if contentLength > 0  ⇒ HttpEntity.Default(contentType, contentLength, data)
+            case ByteHttp2SubStream(_, data)                       ⇒ HttpEntity.Chunked.fromData(contentType, data)
```

Comment:
Maybe add a comment `/* contentLength undefined */` so it's clear why this case may happen.